File: src/test/java/calculator/CalculatorTest.java
package calculator;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class CalculatorTest {

    private final Calculator calculator = new Calculator();

    @DisplayName("문자열이 숫자하나만 들어왔을 경우 숫자를 반환한다.")
    @ValueSource(strings = {"1", "22", "333"})
    @ParameterizedTest
    void onlyNumberCheck(String input) {
        assertThat(calculator.calculate(input)).isEqualTo(Integer.parseInt(input));
    }

    @DisplayName("문자열이 하나만 들어왔을 경우 숫자가 아니라면 예외를 발생시킨다.")
    @ValueSource(strings = {"a", "aa", "a1a"})
    @ParameterizedTest
    void onlyNumberExceptionCheck(String input) {
        assertThatThrownBy(() -> calculator.calculate(input))
                .isInstanceOf(RuntimeException.class)
                .hasMessage("숫자의 형태가 아닙니다.");
    }

    @DisplayName("음수를 전달할 경우 IllegalArgumentException 예외가 발생해야 한다.")
    @ValueSource(strings = {"-1,,1", "1,-2:3", "//;\\n1;;2;;-3"})
    @ParameterizedTest
    void minusNumberCheck(String input) {
        assertThatThrownBy(() -> calculator.calculate(input))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("음수는 들어올 수 없습니다.");
    }

}

File: src/test/kotlin/calculator/CalculatorTest.kt
package calculator

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatRuntimeException
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.junit.jupiter.params.provider.NullAndEmptySource
import org.junit.jupiter.params.provider.ValueSource

class CalculatorTest {

    @ParameterizedTest
    @CsvSource("0,1,2|3", "1,2,3|6", delimiter = '|')
    fun `기본 구분자와 숫자 계산 성공`(target: String, expected: Int) {
        assertThat(SUT.calculate(target)).isEqualTo(expected)
    }

    @ParameterizedTest
    @CsvSource("""//;\n0;2;3|5""", """//;\n1;0;3|4""", """//.\n1.0.3|4""", delimiter = '|')
    fun `커스텀 구분자와 숫자 계산 성공`(target: String, expected: Int) {
        assertThat(SUT.calculate(target)).isEqualTo(expected)
    }

    @ParameterizedTest
    @ValueSource(strings = ["""//;\n0;2,3""", """//;\n1:0;3"""])
    fun `커스텀 구분자와 기본 구분자를 섞어 사용할 수 없음`(target: String) {
        assertThatRuntimeException().isThrownBy { SUT.calculate(target) }
    }

    @ParameterizedTest
    @NullAndEmptySource
    fun `빈 문자열 또는 널`(target: String?) {
        assertThat(SUT.calculate(target)).isEqualTo(0)
    }

    @Test
    fun `기본 구분자와 숫자가 아닌 문자`() {
        val target = "a,b,c"
        assertThatRuntimeException().isThrownBy { SUT.calculate(target) }
    }

    @Test
    fun `기본 구분자와 음수`() {
        val target = "-1,2,3"
        assertThatRuntimeException().isThrownBy { InputString.of(target) }
    }

    companion object {
        private val SUT = Calculator()
    }
}
